pkg/manager/volumes/vol_compare_utils.go

```diff
@@ -0,0 +1,385 @@
+		Context: context.TODO(),
+		tc:      tc,
+		status:  status,
+	}
+
+	vs, err := u.GetDesiredVolumes(tc, comp)
+	if err != nil {
+		return nil, err
+	}
+	ctx.desiredVolumes = vs
+
+	sts, err := u.getStsOfComponent(tc, comp)
+	if err != nil {
+		return nil, err
+	}
+
+	pods, err := u.getPodsOfComponent(tc, comp)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx.pods = pods
+	ctx.sts = sts
+	ctx.shouldEvict = comp == v1alpha1.TiKVMemberType
+
+	return ctx, nil
+}
+
+func (u *volCompareUtils) getPodsOfComponent(tc *v1alpha1.TidbCluster, mt v1alpha1.MemberType) ([]*corev1.Pod, error) {
+	selector, err := u.sf.NewSelector(tc.GetInstanceName(), mt)
+	if err != nil {
+		return nil, err
+	}
+
+	ns := tc.GetNamespace()
+
+	pods, err := u.deps.PodLister.Pods(ns).List(selector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list Pods: %w", err)
+	}
+
+	sort.Slice(pods, func(i, k int) bool {
+		a, b := pods[i].Name, pods[k].Name
+		if len(a) != len(b) {
+			return len(a) < len(b)
+		}
+		return a < b
+	})
+
+	return pods, nil
+}
+
+func (u *volCompareUtils) getStsOfComponent(cluster v1alpha1.Cluster, mt v1alpha1.MemberType) (*appsv1.StatefulSet, error) {
+	ns := cluster.GetNamespace()
+	stsName := controller.MemberName(cluster.GetName(), mt)
+
+	sts, err := u.deps.StatefulSetLister.StatefulSets(ns).Get(stsName)
+	if err != nil {
+		return nil, fmt.Errorf("get sts %s/%s failed: %w", ns, stsName, err)
+	}
+
+	return sts, nil
+}
+
+func (u *volCompareUtils) IsStatefulSetSynced(ctx *componentVolumeContext, sts *appsv1.StatefulSet, errOnNewVol bool) (bool, error) {
+	var newVolErr error = nil
+	for _, volTemplate := range sts.Spec.VolumeClaimTemplates {
+		volName := v1alpha1.StorageVolumeName(volTemplate.Name)
+		size := getStorageSize(volTemplate.Spec.Resources.Requests)
+		desired := getDesiredVolumeByName(ctx.desiredVolumes, volName)
+		if desired == nil {
+			// TODO: verify behaviour is correctly intended, this was previously being ignored. Now explicit error in
+			// pvc modifier..
+			errStr := fmt.Sprintf("volume %s in sts for cluster %s does not exist in desired volumes", volName, ctx.ComponentID())
+			if errOnNewVol {
+				newVolErr = errors.New(errStr)
+			} else {
+				klog.Warningf(errStr)
+			}
+			return false, newVolErr
+		}
+		if size.Cmp(desired.Size) != 0 {
+			return false, nil
+		}
+		scName := volTemplate.Spec.StorageClassName
+		if !isStorageClassMatched(desired.StorageClass, ignoreNil(scName)) {
+			return false, nil
+		}
+	}
+	// TODO: verify new behaviour is fine with pvc_modifier, previously number of volumes not checked.
+	if len(sts.Spec.VolumeClaimTemplates) != len(ctx.desiredVolumes) {
+		errStr := fmt.Sprintf("Number of volumes mismatch desired %d vs sts %d for cluster %s", len(ctx.desiredVolumes), len(sts.Spec.VolumeClaimTemplates), ctx.ComponentID())
+		if errOnNewVol {
+			newVolErr = errors.New(errStr)
+		} else {
+			klog.Warningf(errStr)
+		}
+		return false, newVolErr
+	}
+
+	return true, nil
```

Comment:
It seems in this function, the returned error is only `newVolErr`. I think instead of passing in `errOnNewVol` as a flag to differentiate the behavior on emiting errors or not, maybe it's simpler to let the caller of this function to decide whether to propagate the error or not. That said, we can do `isSynced, err := p.utils.IsStatefulSetSynced(ctx, ctx.sts)` if the error is needed or `isSynced, _ := p.utils.IsStatefulSetSynced(ctx, ctx.sts)` if error can be ignored, what do you think?

Reply:
Done. Also added comments in places where it is ignored.